File: leetcode_learning/code_cpp/63.cpp
#include<iostream>
#include<vector>

using namespace std;

//dp[i][j]表示从(0,0)到(i,j)的路径数
// class Solution {
// public:
//     int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
//         int m = obstacleGrid.size();
//         int n = obstacleGrid[0].size();
//         vector<vector<int>> dp(m, vector<int>(n,0));
//         //边界条件,如果其中有一个障碍，则后续路径数均为0
//         for(int i=0;i<m;i++)
//         {
//             if(obstacleGrid[i][0]==0)
//                 dp[i][0] = 1;
//             else
//                 break;
//         }
//         for(int j=0;j<n;j++)
//         {
//             if(obstacleGrid[0][j]==0)
//                 dp[0][j] = 1;
//             else
//                 break;
//         }

//         for(int i=1;i<m;i++)
//             for(int j=1;j<n;j++)
//             {
//                 if(obstacleGrid[i][j]==1)
//                     dp[i][j] = 0;
//                 else
//                 {
//                     dp[i][j] = dp[i-1][j] + dp[i][j-1];
//                 }
//             }
        
//         return dp[m-1][n-1];
//     }
// };


//递推只涉及到相邻两行或两列的状态，可以使用滚动数组优化
//dp_row表示dp每一行的状态
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        //vector<vector<int>> dp(m, vector<int>(n,0));
        vector<int> dp_row(n,0);
        dp_row[0] = (obstacleGrid[0][0]==0);

        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
            {
                if(obstacleGrid[i][j]==1)
                    dp_row[j] = 0;
                if(j>0&&obstacleGrid[i][j]==0)
                {
                    dp_row[j] += dp_row[j-1];
                }
            }
        
        return dp_row[n-1];
    }
};

File: leetcode_learning/auto-leetcode/63.不同路径II/63.py
"""
63.不同路径II
难度：中等
链接：https://leetcode.cn/problems/unique-paths-ii/
"""
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        # dp[i][j] 表示从左上角到 (i,j) 的不同路径数
        # 障碍物处路径数为 0，非障碍物处 dp[i][j] = dp[i-1][j] + dp[i][j-1]
        n, m = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0] * m for _ in range(n)]

        # 起点：非障碍物则路径数为 1，障碍物则路径数为 0（整个网格无解）
        dp[0][0] = 1 if obstacleGrid[0][0] == 0 else 0

        # 第一列：只能从上方到达，遇到障碍物则该位置及下方都为 0
        for i in range(1, n):
            dp[i][0] = dp[i - 1][0] if obstacleGrid[i][0] == 0 else 0

        # 第一行：只能从左边到达，遇到障碍物则该位置及右方都为 0
        for j in range(1, m):
            dp[0][j] = dp[0][j - 1] if obstacleGrid[0][j] == 0 else 0

        # 其余位置：非障碍物处从上方和左方路径数相加
        for i in range(1, n):
            for j in range(1, m):
                if obstacleGrid[i][j] == 0:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
                # 障碍物处 dp[i][j] 保持初始值 0，无法通过

        return dp[n - 1][m - 1]